classes/question_db_helper.php

```diff
@@ -0,0 +1,193 @@
+            }
+
+            if (count($update) > 1) {
+                $DB->update_record(self::QUESTION_TABLE, (object)$update);
+            }
+        } else {
+            // Insert a new record with the question state only containing the options.
+            $state = $this->options_to_state($question, []);
+            $DB->insert_record(self::QUESTION_TABLE, [
+                "questionid" => $question->id,
+                "feedback" => "",
+                "packageid" => $packageid,
+                "state" => json_encode($state),
+            ]);
+        }
+    }
+
+    /**
+     * Deletes all QuestionPy-specific data for the given question.
+     *
+     * @param int $questionid
+     * @throws dml_exception
+     */
+    public function delete_question(int $questionid) {
+        global $DB;
+        $DB->delete_records(self::QUESTION_TABLE, ['questionid' => $questionid]);
+        // TODO: Also delete packages when they are no longer used by any question.
+    }
+
+    /**
+     * Updates the state to contain all package-specific form options in the given question or creates a new state.
+     *
+     * TODO: Eventually, the question state will be an opaque string or JSON string whose content we don't directly
+     * access, but which we send to the package and which it uses to generate the new form definition for us. For now,
+     * however, we assume that the state contains the options and access them directly.
+     *
+     * @param object $question question data, NOT an instance of {@see \question_definition}
+     * @param array $state     previous state of the question
+     * @return array new state of the question, with the form options added or updated
+     */
+    private function options_to_state(object $question, array $state): array {
+        foreach ($question as $key => $value) {
+            $unmangled = form_name_mangler::unmangle($key);
+            if (!$unmangled) {
+                // Not a package-specific form option.
+                continue;
+            }
+
+            $state[$unmangled] = $value;
+        }
+
+        return $state;
+    }
+
+    /**
+     * Get the package with the given hash from the DB or the QuestionPy server API.
+     *
+     * If the package isn't found in the DB, then it is retrieved from the API and stored in the DB. If it isn't found
+     * by the API either, `null` is returned.
+     *
+     * @param string $hash hash of the package to look for
+     * @return ?array [id of database record, package instance] or null if package not found in DB or API
+     * @throws dml_exception
+     * @throws moodle_exception
+     */
+    private function get_package(string $hash): ?array {
+        $result = package::get_record_by_hash($hash);
+        if ($result) {
+            return $result;
+        }
+
+        $package = $this->api->get_package($hash);
+        if (!$package) {
+            return null;
+        }
+        return [$package->store_in_db(), $package];
+    }
```

Comment:
Vielleicht könnte man die Methode in eine passendere Klasse wie z.B. `db_helper` verschieben, da sie ja nicht direkt mit den einzelnen Fragen im Zusammenhang steht? Alternativ könnte man auch direkt die gesamte Klasse umbenennen.

Reply:
Ah, hab jetzt gerade deinen Kommentar zum Klassennamen gesehen..

Reply:
Da es aktuell nur hier benötigt wird, würde ich das erstmal so lassen. Sollte es noch woanders nötig werden, können wir es in eine dritte Klasse verschieben.